pkg/securelaunch/policy/policy.go

```diff
@@ -0,0 +1,227 @@
+package policy
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+
+	"github.com/u-root/u-root/pkg/cmdline"
+	"github.com/u-root/u-root/pkg/diskboot"
+	"github.com/u-root/u-root/pkg/find"
+	"github.com/u-root/u-root/pkg/mount"
+	slaunch "github.com/u-root/u-root/pkg/securelaunch"
+	"github.com/u-root/u-root/pkg/securelaunch/eventlog"
+	"github.com/u-root/u-root/pkg/securelaunch/launcher"
+	"github.com/u-root/u-root/pkg/securelaunch/measurement"
+)
+
+/*
+ * policy structure passed around various sub-packages (
+ * measurement, eventlog and launcher) in securelaunch.
+ * Used to store parsed policy file. The policy file
+ * is a JSON file.
+ */
+type policy struct {
+	DefaultAction string
+	Collectors    []measurement.Collector
+	Launcher      launcher.Launcher
+	EventLog      eventlog.EventLog
+}
+
+/*
+ * scanKernelCmdLine() ([]byte, error) scans the kernel cmdline
+ * for 'sl_policy' flag. when set, this flag provides location of
+ * of policy file on disk.
+ *
+ * format of sl_policy flag is as follows
+ * sl_policy=<block device identifier>:<path>
+ * e.g sda:/boot/securelaunch.policy
+ * e.g 4qccd342-12zr-4e99-9ze7-1234cb1234c4:/foo/securelaunch.policy
+ */
+func scanKernelCmdLine() ([]byte, error) {
+
+	slaunch.Debug("scanKernelCmdLine: scanning kernel cmd line for *sl_policy* flag")
+	val, ok := cmdline.Flag("sl_policy")
+	if !ok {
+		log.Printf("scanKernelCmdLine: sl_policy cmdline flag is not set")
+		return nil, errors.New("sl_policy flag not set")
+	}
+
+	// val is of type sda:path
+	mntFilePath, mountPath, e := slaunch.GetMountedFilePath(val, false) // false means readonly mount
+	if e != nil {
+		log.Printf("scanKernelCmdLine: GetMountedFilePath err=%v", e)
+		return nil, fmt.Errorf("scanKernelCmdLine: GetMountedFilePath err=%v", e)
+	}
+	slaunch.Debug("scanKernelCmdLine: Reading file=%s", mntFilePath)
+
+	d, err := ioutil.ReadFile(mntFilePath)
+	if e := mount.Unmount(mountPath, true, false); e != nil {
+		log.Printf("Unmount failed. PANIC")
+		panic(e)
+	}
+
+	if err != nil {
+		// - TODO: should we check for end of file ?
+		log.Printf("Error reading policy file:mountPath=%s, passed=%s", mntFilePath, val)
+		return nil, fmt.Errorf("failed to read policy file:mountPath=%s, passed=%s", mntFilePath, val)
+	}
+	return d, nil
+}
+
+/*
+ *  scanBlockDevice recursively scans an already mounted block device inside directories
+ *	"/", "/efi" and "/boot" for policy file and if found, returns the policy byte as a byte slice.
+ *
+ *	e.g: if you mount /dev/sda1 on /tmp/sda1,
+ *	then mountPath would be /tmp/sda1
+ *	and searchPath would be /tmp/sda1/, /tmp/sda1/efi, and /tmp/sda1/boot
+ *		respectively for each iteration of loop over SearchRoots slice.
+ */
+func scanBlockDevice(mountPath string) ([]byte, bool) {
+
+	log.Printf("scanBlockDevice")
+	// scan for securelaunch.policy under /, /efi, or /boot
+	var SearchRoots = []string{"/", "/efi", "/boot"}
```

Comment:
Thats an excellent point which made us think we dont need a recursive function to search for policy file. We actually wantt to just search for max_depth = 1 since in production typically a default/samply policy file is kept in /usr/share ,and we dont want that policy file to be read instead....

So i will remove the find call and just check using os.Stat for the three locations...
/securelaunch.policy
/boot/securelaunch.policy
/efi/securelaunch.policy.
Again thanks for this. if you hadn't made this comment, we would have realzed this later only to come back and change it